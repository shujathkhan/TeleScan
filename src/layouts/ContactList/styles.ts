import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  container: {
    flexGrow: 1,
    flexDirection: 'column',
    backgroundColor: '#FFFBDF',
    minHeight: 70,
  },
  fallbackStatusView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  fallbackStatusText: {
    fontSize: 15,
  },
  fallbackStatusTextView: { marginVertical: 20 },
  infoContainer: {
    padding: 10,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    backgroundColor: 'white',
    elevation: 3,
    marginBottom: 5,
  },
  infoText: { fontSize: 15 },
  typeOfView: {
    flexDirection: 'row',
    borderRadius: 5,
    width: 80,
    paddingVertical: 5,
    paddingHorizontal: 10,
    backgroundColor: '#61dafbb0',
    alignItems: 'center',
  },
  separator: {
    borderWidth: 0.75,
    height: 25,
    marginHorizontal: 10,
  },
  fabView: {
    position: 'absolute',
    bottom: '7%',
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'flex-end',
  },
  deleteView: {
    backgroundColor: 'red',
    justifyContent: 'center',
    alignItems: 'flex-end',
    paddingHorizontal: 30,
    marginHorizontal: 7.5,
    paddingVertical: 5,
    marginVertical: 10,
    borderRadius: 10,
  },
  deleteText: {
    color: 'white',
    fontSize: 20,
    fontWeight: '600',
  },
});
