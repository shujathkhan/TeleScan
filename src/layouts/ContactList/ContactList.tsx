import { NativeStackNavigationHelpers } from '@react-navigation/native-stack/lib/typescript/src/types';
import React, { useEffect, useState } from 'react';
import { FlatList, RefreshControl, Text, View, Image, SafeAreaView, TouchableOpacity, ListRenderItem } from 'react-native';
import Contacts, { Contact } from 'react-native-contacts';

import { contactPermissions, requestPermissions, sortByGivenName, wait } from '../../helpers';
import ContactCard from '../../components/ContactCard';
import FabButton from '../../components/FabButton';
import { styles } from './styles';
import Snackbar from 'react-native-snackbar';
import Swipeable from 'react-native-gesture-handler/Swipeable';

type TContactList = {
  navigation: NativeStackNavigationHelpers | any;
};

type TFlatListItem = { item: Contact; index: number };

/**
 * @description All Contacts screen component
 * @param {TContactList} props
 * @returns {any}
 */
const ContactList = (props: TContactList) => {
  const [contactList, setContactList] = useState<Array<Contact>>([]);
  const [refreshing, setRefreshing] = useState<boolean>(false);

  /**
   * @description Function to call getAll API
   * @returns {void}
   */
  const loadContacts = () => {
    Contacts.getAll()
      .then(contacts => {
        let sortedContact = sortByGivenName(contacts);
        setContactList(sortedContact);
        Snackbar.show({
          text: `Successfully loaded ${contacts.length} contacts 💯`,
          duration: Snackbar.LENGTH_SHORT,
          backgroundColor: '#29BB89',
          textColor: 'white',
        });
      })
      .catch(err => {
        setContactList([]);
        console.error(err);
        Snackbar.show({
          text: 'Oops, we have a situation! 🚧',
          duration: Snackbar.LENGTH_SHORT,
          backgroundColor: 'red',
          textColor: 'white',
        });
      });
  };

  /**
   * @description Function to request access permissions
   * @returns {void}
   */
  const handleRequestAccess = async () => {
    const isPermissionGranted = await requestPermissions(contactPermissions);
    if (isPermissionGranted) {
      loadContacts();
    } else {
      setContactList([]);
      Snackbar.show({
        text: 'Well, feel free to explore 👍',
        duration: Snackbar.LENGTH_SHORT,
        backgroundColor: '#185ADB',
        textColor: 'white',
      });
    }
  };

  useEffect(() => {
    handleRequestAccess();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  /**
   * @description Function to handle delete button press event
   * @param {TFlatListItem} childProps
   * @returns {void}
   */
  const handleDelete = (childProps: TFlatListItem) => {
    const contactName = childProps.item.displayName;
    Contacts.deleteContact(childProps.item)
      .then(() => {
        loadContacts();
        Snackbar.show({
          text: `Successfully deleted ${contactName} ✔️`,
          duration: Snackbar.LENGTH_SHORT,
          backgroundColor: '#29BB89',
          textColor: 'white',
        });
      })
      .catch(err => {
        console.error(err);
        Snackbar.show({
          text: 'Oops, we have a situation! 🚧',
          duration: Snackbar.LENGTH_SHORT,
          backgroundColor: 'red',
          textColor: 'white',
        });
      });
  };

  /**
   * @description List item component to be passed to Flat List
   * @param {TFlatListItem} childProps
   * @returns {ListRenderItem<Contact>}
   */
  const contactItem: ListRenderItem<Contact> = (childProps: TFlatListItem) => {
    /**
     * @description Render on swipe right delete button
     * @returns {JSX.Element}
     */
    const handleRightSwipe = () => (
      <TouchableOpacity style={styles.deleteView} activeOpacity={0.75} onPress={() => handleDelete(childProps)}>
        <Text style={styles.deleteText}>Delete</Text>
      </TouchableOpacity>
    );

    return childProps.item.displayName && childProps.item.displayName !== 'undefined' ? (
      <Swipeable renderRightActions={handleRightSwipe} onSwipeableLeftOpen={() => console.log('Swiped')}>
        <ContactCard
          name={childProps.item.displayName}
          nameIconPath={childProps.item.thumbnailPath}
          onPress={() =>
            props.navigation.navigate('ContactView', {
              contactId: childProps.item.recordID,
            })
          }
          key={childProps.item.recordID}
        />
      </Swipeable>
    ) : null;
  };

  /**
   * @description onRefresh function passed to FlatList
   * @param {any} (
   * @returns {React.EffectCallback}
   */
  const onRefresh: React.EffectCallback = React.useCallback(() => {
    setRefreshing(true);
    loadContacts();
    wait(250).then(() => setRefreshing(false));
  }, []);

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.infoContainer}>
        <Text style={styles.infoText}>Showing {contactList.length} contacts</Text>
      </View>

      {contactList.length ? (
        <>
          <FlatList
            testID="flat-list"
            data={contactList}
            renderItem={contactItem}
            initialNumToRender={10}
            keyExtractor={item => item.recordID}
            refreshControl={<RefreshControl refreshing={refreshing} onRefresh={onRefresh} />}
          />

          <View
            style={styles.fabView}
            accessibilityLabel="Add new contact"
            testID="addButton"
            accessible
            accessibilityHint="Navigates to new contact screen">
            <FabButton
              type="add"
              onPress={() => {
                props.navigation.navigate('NewContact');
              }}
            />
          </View>
        </>
      ) : (
        <View style={styles.fallbackStatusView}>
          <Image source={require('../../assets/blank.png')} />
          <View style={styles.fallbackStatusTextView}>
            <Text style={styles.fallbackStatusText}>No contacts to display</Text>
            <TouchableOpacity activeOpacity={0.75} onPress={handleRequestAccess} testID="syncContacts">
              <Text style={styles.syncText}>Sync device contacts</Text>
            </TouchableOpacity>
          </View>
        </View>
      )}
    </SafeAreaView>
  );
};

export default ContactList;
