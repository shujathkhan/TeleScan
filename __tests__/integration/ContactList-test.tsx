import React from 'react';
import { render } from '@testing-library/react-native';
import ContactList from '../../src/layouts/ContactList';

describe('All Contacts Screen', () => {
  const navigation = {
    navigate: jest.fn(),
    setOptions: jest.fn(),
    goBack: jest.fn(),
  };
  it('should render ContactList screen', async () => {
    const tree = render(<ContactList navigation={navigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
