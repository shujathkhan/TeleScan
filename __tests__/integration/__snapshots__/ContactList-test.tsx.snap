// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Contacts Screen should render ContactList screen 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#FFFBDF",
      "flexDirection": "column",
      "flexGrow": 1,
      "minHeight": 70,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "elevation": 3,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 5,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 15,
        }
      }
    >
      Showing 
      0
       contacts
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "process": [Function],
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginVertical": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 15,
          }
        }
      >
        No contacts to display
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="syncContacts"
      >
        <Text
          style={
            Object {
              "color": "blue",
              "fontSize": 15,
              "fontWeight": "700",
            }
          }
        >
          Sync device contacts
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
